Body of messages not fetched

When using one of my accounts, some messages (which heavily use html) are not correctly fetched. I can see the title and some headers but the body is missing.

This happens only with one mail account.

The cause is probably this exception:
```
11-14 07:55:17.157 23585 23602 E k9      : synchronizeMailbox
11-14 07:55:17.157 23585 23602 E k9      : com.fsck.k9.mailstore.LockableDatabase$WrappedException: com.fsck.k9.mail.MessagingException: Error appending message: debian-devel-digest Digest V2016 #603
11-14 07:55:17.157 23585 23602 E k9      : 	at com.fsck.k9.mailstore.LocalFolder$12.doDbWork(LocalFolder.java:1090)
11-14 07:55:17.157 23585 23602 E k9      : 	at com.fsck.k9.mailstore.LocalFolder$12.doDbWork(LocalFolder.java:1078)
11-14 07:55:17.157 23585 23602 E k9      : 	at com.fsck.k9.mailstore.LockableDatabase.execute(LockableDatabase.java:275)
11-14 07:55:17.157 23585 23602 E k9      : 	at com.fsck.k9.mailstore.LocalFolder.storeSmallMessage(LocalFolder.java:1078)
11-14 07:55:17.157 23585 23602 E k9      : 	at com.fsck.k9.controller.MessagingController$12.messageFinished(MessagingController.java:1406)
11-14 07:55:17.157 23585 23602 E k9      : 	at com.fsck.k9.mail.store.imap.ImapFolder.fetch(ImapFolder.java:770)
11-14 07:55:17.157 23585 23602 E k9      : 	at com.fsck.k9.controller.MessagingController.downloadSmallMessages(MessagingController.java:1393)
11-14 07:55:17.157 23585 23602 E k9      : 	at com.fsck.k9.controller.MessagingController.downloadMessages(MessagingController.java:1196)
11-14 07:55:17.157 23585 23602 E k9      : 	at com.fsck.k9.controller.MessagingController.synchronizeMailboxSynchronous(MessagingController.java:953)
11-14 07:55:17.157 23585 23602 E k9      : 	at com.fsck.k9.controller.MessagingController$9.run(MessagingController.java:739)
11-14 07:55:17.157 23585 23602 E k9      : 	at com.fsck.k9.controller.MessagingController.runInBackground(MessagingController.java:204)
11-14 07:55:17.157 23585 23602 E k9      : 	at com.fsck.k9.controller.MessagingController.access$000(MessagingController.java:115)
11-14 07:55:17.157 23585 23602 E k9      : 	at com.fsck.k9.controller.MessagingController$1.run(MessagingController.java:173)
11-14 07:55:17.157 23585 23602 E k9      : 	at java.lang.Thread.run(Thread.java:818)
11-14 07:55:17.157 23585 23602 E k9      : Caused by: com.fsck.k9.mail.MessagingException: Error appending message: debian-devel-digest Digest V2016 #603
11-14 07:55:17.157 23585 23602 E k9      : 	at com.fsck.k9.mailstore.LocalFolder.saveMessage(LocalFolder.java:1326)
11-14 07:55:17.157 23585 23602 E k9      : 	at com.fsck.k9.mailstore.LocalFolder$14.doDbWork(LocalFolder.java:1187)
11-14 07:55:17.157 23585 23602 E k9      : 	at com.fsck.k9.mailstore.LocalFolder$14.doDbWork(LocalFolder.java:1182)
11-14 07:55:17.157 23585 23602 E k9      : 	at com.fsck.k9.mailstore.LockableDatabase.execute(LockableDatabase.java:275)
11-14 07:55:17.157 23585 23602 E k9      : 	at com.fsck.k9.mailstore.LocalFolder.appendMessages(LocalFolder.java:1182)
11-14 07:55:17.157 23585 23602 E k9      : 	at com.fsck.k9.mailstore.LocalFolder.appendMessages(LocalFolder.java:1109)
11-14 07:55:17.157 23585 23602 E k9      : 	at com.fsck.k9.mailstore.LocalFolder$12.doDbWork(LocalFolder.java:1082)
11-14 07:55:17.157 23585 23602 E k9      : 	... 13 more
11-14 07:55:17.157 23585 23602 E k9      : Caused by: java.lang.NullPointerException: Attempt to invoke virtual method 'void java.io.InputStream.close()' on a null object reference
11-14 07:55:17.157 23585 23602 E k9      : 	at com.fsck.k9.mail.internet.MimeUtility.closeInputStreamWithoutDeletingTemporaryFiles(MimeUtility.java:1060)
11-14 07:55:17.157 23585 23602 E k9      : 	at com.fsck.k9.mail.internet.MessageExtractor.getTextFromTextPart(MessageExtractor.java:112)
11-14 07:55:17.157 23585 23602 E k9      : 	at com.fsck.k9.mail.internet.MessageExtractor.getTextFromPart(MessageExtractor.java:55)
11-14 07:55:17.157 23585 23602 E k9      : 	at com.fsck.k9.crypto.MessageDecryptVerifier.isPartPgpInlineEncrypted(MessageDecryptVerifier.java:207)
11-14 07:55:17.157 23585 23602 E k9      : 	at com.fsck.k9.message.extractors.EncryptionDetector.containsInlinePgpEncryptedText(EncryptionDetector.java:34)
11-14 07:55:17.157 23585 23602 E k9      : 	at com.fsck.k9.message.extractors.EncryptionDetector.isEncrypted(EncryptionDetector.java:25)
11-14 07:55:17.157 23585 23602 E k9      : 	at com.fsck.k9.message.extractors.MessagePreviewCreator.createPreview(MessagePreviewCreator.java:31)
11-14 07:55:17.157 23585 23602 E k9      : 	at com.fsck.k9.mailstore.LocalFolder.saveMessage(LocalFolder.java:1251)
11-14 07:55:17.157 23585 23602 E k9      : 	... 19 more
11-14 07:55:17.173 23585 23602 E k9      : Failed synchronizing folder spamowe:INBOX @ Mon Nov 14 07:55:17 GMT+01:00 2016
```

### Environment
K-9 Mail version: 5.114

Android version: CM 13.0

Account type (IMAP, POP3, WebDAV/Exchange): IMAP

--- a/k9mail-library/src/main/java/com/fsck/k9/mail/internet/MimeBodyPart.java
+++ b/k9mail-library/src/main/java/com/fsck/k9/mail/internet/MimeBodyPart.java
@@ -4,6 +4,7 @@ package com.fsck.k9.mail.internet;
 import com.fsck.k9.mail.Body;
 import com.fsck.k9.mail.BodyPart;
 import com.fsck.k9.mail.MessagingException;
+import com.fsck.k9.mail.Multipart;
 
 import java.io.BufferedWriter;
 import java.io.IOException;
@@ -93,7 +94,14 @@ public class MimeBodyPart extends BodyPart {
     @Override
     public String getContentType() {
         String contentType = getFirstHeader(MimeHeader.HEADER_CONTENT_TYPE);
-        return (contentType == null) ? "text/plain" : MimeUtility.unfoldAndDecode(contentType);
+        if (contentType != null) {
+            return MimeUtility.unfoldAndDecode(contentType);
+        }
+        Multipart parent = getParent();
+        if (parent != null && "multipart/digest".equals(parent.getMimeType())) {
+            return "message/rfc822";
+        }
+        return "text/plain";
     }
 
     @Override

--- a/k9mail/src/test/java/com/fsck/k9/mailstore/MessageViewInfoExtractorTest.java
+++ b/k9mail/src/test/java/com/fsck/k9/mailstore/MessageViewInfoExtractorTest.java
@@ -1,6 +1,7 @@
 package com.fsck.k9.mailstore;
 
 
+import java.io.ByteArrayInputStream;
 import java.util.ArrayList;
 import java.util.Date;
 import java.util.List;
@@ -25,6 +26,7 @@ import com.fsck.k9.mail.internet.MimeMessageHelper;
 import com.fsck.k9.mail.internet.MimeMultipart;
 import com.fsck.k9.mail.internet.TextBody;
 import com.fsck.k9.mail.internet.Viewable;
+import com.fsck.k9.mail.internet.Viewable.MessageHeader;
 import com.fsck.k9.mailstore.MessageViewInfoExtractor.ViewableExtractedText;
 import org.junit.Before;
 import org.junit.Test;
@@ -264,6 +266,71 @@ public class MessageViewInfoExtractorTest {
 
         assertEquals(expectedText, container.text);
         assertEquals(expectedHtml, getHtmlBodyText(container.html));
+    }
+
+    @Test
+    public void testMultipartDigestWithMessages() throws Exception {
+        String data = "Content-Type: multipart/digest; boundary=\"bndry\"\r\n" +
+                "\r\n" +
+                "--bndry\r\n" +
+                "\r\n" +
+                "Content-Type: text/plain\r\n" +
+                "\r\n" +
+                "text body of first message\r\n" +
+                "\r\n" +
+                "--bndry\r\n" +
+                "\r\n" +
+                "Subject: subject of second message\r\n" +
+                "Content-Type: multipart/alternative; boundary=\"bndry2\"\r\n" +
+                "\r\n" +
+                "--bndry2\r\n" +
+                "Content-Type: text/plain\r\n" +
+                "\r\n" +
+                "text part of second message\r\n" +
+                "\r\n" +
+                "--bndry2\r\n" +
+                "Content-Type: text/html\"\r\n" +
+                "\r\n" +
+                "html part of second message\r\n" +
+                "\r\n" +
+                "--bndry2--\r\n" +
+                "\r\n" +
+                "--bndry--\r\n";
+        MimeMessage message = MimeMessage.parseMimeMessage(new ByteArrayInputStream(data.getBytes()), false);
+
+        // Extract text
+        List<Part> outputNonViewableParts = new ArrayList<>();
+        ArrayList<Viewable> outputViewableParts = new ArrayList<>();
+        MessageExtractor.findViewablesAndAttachments(message, outputViewableParts, outputNonViewableParts);
+
+        String expectedExtractedText = "Subject: (No subject)\r\n" +
+                "\r\n" +
+                "text body of first message\r\n" +
+                "\r\n" +
+                "\r\n" +
+                "------------------------------------------------------------------------\r\n" +
+                "\r\n" +
+                "Subject: subject of second message\r\n" +
+                "\r\n" +
+                "text part of second message\r\n";
+        String expectedHtmlText = "<table style=\"border: 0\">" +
+                "<tr><th style=\"text-align: left; vertical-align: top;\">Subject:</th><td>(No subject)</td></tr>" +
+                "</table>" +
+                "<pre class=\"k9mail\">text body of first message<br /></pre>" +
+                "<p style=\"margin-top: 2.5em; margin-bottom: 1em; border-bottom: 1px solid #000\"></p>" +
+                "<table style=\"border: 0\">" +
+                "<tr><th style=\"text-align: left; vertical-align: top;\">Subject:</th><td>subject of second message</td></tr>" +
+                "</table>" +
+                "<pre class=\"k9mail\">text part of second message<br /></pre>";
+
+
+        assertEquals(4, outputViewableParts.size());
+        assertEquals("subject of second message", ((MessageHeader) outputViewableParts.get(2)).getMessage().getSubject());
+
+        ViewableExtractedText firstMessageExtractedText =
+                messageViewInfoExtractor.extractTextFromViewables(outputViewableParts);
+        assertEquals(expectedExtractedText, firstMessageExtractedText.text);
+        assertEquals(expectedHtmlText, getHtmlBodyText(firstMessageExtractedText.html));
     }
 
     private static String getHtmlBodyText(String htmlText) {

